### DISCUSSION TITLE
Constraint not being followed [Resolved]

### DISCUSSION THREAD

**Discussion by Sorah-Darkhat** [MilesCranmer/PySR#717] [Q&A] [answered]
Hello!

I’ve been encountering some issues with PySRRegressor, where the returned functions occasionally produce complex values.

I would like to know if it’s possible to restrict the functions being evaluated to only return real values within the bounds of the x-vector, i.e., for every xi value between the minimum and maximum of x.

Although the imaginary component is usually small, I cannot work with complex numbers.

I’m also inclined to retain the potentially problematic operators, like "pow" and "log".

So far, the only solution I’ve considered is applying constraints on subtraction and addition inside the "pow" and "log" operators. However, I’m not confident this would guarantee real-valued functions, and it might lead to less optimal solutions.

Thank you in advance!

**Comment by MilesCranmer** [comment]
Hi @Sorah-Darkhat,

PySRRegressor will only return complex values if you train it on complex data. Be sure to verify that you are passing real data to avoid this. e.g., you can check `X.dtype` and `y.dtype`.

Cheers,
Miles

**Reply by Sorah-Darkhat** [reply, marked as answer]
Thank you for your answer,

I'm having trouble reproducing the complex error within the domain of the x-vector, so I'm left believing that I was evaluating it outside the domain.

I've also removed the "log" operation, but I've kept the "^" operator, and sometimes PySRRegressor returns functions of the type:

( C1 - x ) ^ C2

Where C1 and C2 are floats, and for values of x > C1, I get complex solutions.

I know it's possible to load constraints, I've added the following constrain:

constrain = { "^" : [1, 3] }, but it doesn't seem to be implementing the constraint.

( I'm loading the model configurations from a json file, and I'm also running with the experimental turbo and bumper ).
This is the json file:

{
    "binary_operators": ["+", "-", "*", "/", "^"],
    "complexity_of_constants": 1,
    "constraints" : {"^": [1, 3]},
    "maxsize": 10,
    "niterations": 300,
    "populations": 30,
    "dimensional_constraint_penalty": 100,
    "model_selection": "best",
    "temp_equation_file": true,
    "early_stop_condition": "f(loss, complexity) = (loss < 0.00001) && (complexity <= 10)",
    "turbo" : true,
    "bumper" : true
}

With the best regards,

Sorah


[Edit1]: I've changed the title of the discussion, and I've noticed someone else had a similar issue back in February, #541 , however I'm not sure if I'm not just implementing the constraint wrong.

[Edit2]: I've attempted to reproduce the problem, and to do so I no longer load the configuration from the json file, this seems to resolve the issue, although I'm 100% sure why loading it from the json file causes problems with the constraints, since all other parameters seem to be working just fine, including the "turbo" and "bumper", regardless, I will no longer be using the json file setup.

Thanks for the great package, and sorry for any time wasted reading this post.
